core/services/eth/log_broadcaster_test.go

```diff
@@ -318,3 +318,61 @@ func TestDecodingLogListener(t *testing.T) {
 	listener.HandleLog(nil, expectedErr)
 	require.Equal(t, err, expectedErr)
 }
+
+func TestLogBroadcaster_ReceivesAllLogsWhenResubscribing(t *testing.T) {
+	store, cleanup := cltest.NewStore(t)
+	defer cleanup()
+
+	const blockHeight uint64 = 0
+
+	ethClient := new(mocks.Client)
+	sub := new(mocks.Subscription)
+
+	chBlock := make(chan struct{})
+	chchRawLogs := make(chan chan<- eth.Log, 1)
+	var unblock struct{}
+
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			chRawLogs := args.Get(1).(chan<- eth.Log)
+			chchRawLogs <- chRawLogs
+			<-chBlock
+			return
+		}).
+		Return(sub, nil).
+		Twice()
+
+	ethClient.On("GetBlockHeight").Return(blockHeight, nil)
+	sub.On("Err").Return(nil)
+	sub.On("Unsubscribe").Return()
+
+	lb := ethsvc.NewLogBroadcaster(ethClient, store.ORM)
+	lb.Start()
+
+	logCount := 0
+	logListener := funcLogListener{
+		fn: func(log interface{}, err error) { logCount++ },
+	}
+	logListener2 := funcLogListener{
+		fn: func(log interface{}, err error) {},
+	}
+
+	lb.Register(common.Address{}, &logListener)
+	chRawLogs1 := <-chchRawLogs
+	chBlock <- unblock
+	go func() {
+		chRawLogs1 <- eth.Log{BlockNumber: 1, Index: 0}
+		chRawLogs1 <- eth.Log{BlockNumber: 1, Index: 1}
+	}()
+
+	lb.Register(common.Address{1}, &logListener2) // trigger resubscription
+	chRawLogs2 := <-chchRawLogs
+	go func() {
+		chRawLogs2 <- eth.Log{BlockNumber: 1, Index: 1} // send overlapping logs
```

Comment:
To mimic what will happen with a real blockchain, you're gonna want to re-send the `Index: 0` log as well here.